### PR TITLE
to add arweave/README + caip2 + caip10 + caip122

### PR DESCRIPTION
**Pull request by ropats16**
Create a new namespace for arweave. 

Add support for:
- caip2 (Chains)
- caip10 (Addresses)
- caip122 (SIWX)

### CODE REVIEW

**Comment by ligi** [comment]
are you in Berlin at Berlin blockchain week? Our next meeting to discuss PRs is in person there - see: https://github.com/ChainAgnostic/Gatherings

**Comment by ropats16** [comment]
> 

I won't be able to join the Berlin blockchain week but I'd like to know the consensus y'all reach and how we can take this further.

File: arweave/caip10.md
Comment (oed) [inline arweave/caip10.md:35]: Can you mention somewhere what format the address is encoded with? Is it base64?

**Review by oed — COMMENTED** [review]
Looks mostly good. Would like @zachferland to have a look at the caip122 spec

File: arweave/caip10.md
Comment (DanMacDonald) [reply arweave/caip10.md:35]: base64Url encoded 😎

File: arweave/caip2.md
Comment (ropats16) [reply arweave/caip2.md:42]: Arweave has only 1 chain currently which has been designated `1` in the ecosystem. If more chains are added, a system for identifying chains would be created at the time and the `arweave` namespace would be updated accordingly.

File: arweave/caip10.md
Comment (ropats16) [reply arweave/caip10.md]: Updated the PR to reflect the format in which the addresses are encoded.

File: arweave/caip122.md
Comment (zachferland) [inline arweave/caip122.md:18]: does a sign in with arweave standard already exist? this references eth eip

File: arweave/caip122.md
Comment (zachferland) [inline arweave/caip122.md:40]: can it additionally include all field also in EIP-4361?

File: arweave/caip122.md
Comment (ropats16) [reply arweave/caip122.md:18]: Arweave has not needed a sign in standard in the past because of the way the ecosystem works. However, with this integration, there is a use for it. So this CAIP is the first version of the standard. Would love some insights on how we can rephrase the rationale based on this information.

File: arweave/caip122.md
Comment (ropats16) [reply arweave/caip122.md:40]: As mentioned in the comment above, this CAIP is a first version of the standard for signing in with arweave and we are only using the EIP-4361 as a reference. Can having only the required fields for now suffice? We can add more fields as the standard becomes refined.

Comment (zachferland) [reply arweave/caip122.md:40]: ok, yeah in this case would include all required and optional fields, as most optional are needed to enable full use and options of CACAO, and we (ceramic) expect some optional fields to enable our use cases (ie resources, expiration), others may as well (ie nonce for authn)

File: arweave/caip122.md
Comment (zachferland) [reply arweave/caip122.md:18]: yup makes sense, could usual similar wording to here - https://github.com/ChainAgnostic/namespaces/blob/main/solana/caip122.md and mention inspired by eip-4316, or reference

**Comment by zachferland** [comment]
Lastly in the signature creation section, would have to describe how to go from that message to signature creation. Ie from message to bytes/encoding and if arweave has a standard signing interface/implementation (or if ambiguous, add detail). Again ref solana one, which provides a sufficient explanation for example, or eth which has existing EIPs which describe signature creation and verification and are referenced instead.

File: arweave/caip122.md
Comment (zachferland) [inline arweave/caip122.md:74]: message above will already be a string, would you then be able to directly sign that after or would you pass a Uint8Array of the string to be signed?

Comment (ropats16) [reply arweave/caip122.md:74]: If the data model is already a JSON serialised string then we can directly pass it to the sign() method, getting the Uint8Array which would be encoded to Base64 URL format to get the final signature which is the format that is needed to be sent over HTTP to the server for verification.

Comment (ropats16) [reply arweave/caip122.md]: @zachferland updated the PR to reflect the data model being a string.

Comment (zachferland) [reply arweave/caip122.md]: thanks! im good with the caip122 part 👍

**Review by zachferland — APPROVED** [review]

File: arweave/caip122.md
Comment (ropats16) [reply arweave/caip122.md]: @oed if CAIP122 was the last part we needed to get sorted, can we get this merged so that we can move onto the next steps?

**Comment by ropats16** [comment]
> @oed if you have any more feedback and/or want to approve

@oed following up on this to check if it is good to be merged?

File: arweave/caip122.md
Comment (oed) [inline arweave/caip122.md:28]: You need to specify a singature type string. See: https://github.com/ChainAgnostic/CAIPs/blob/master/CAIPs/caip-74.md#signature-verification

File: arweave/caip2.md
Comment (oed) [reply arweave/caip2.md:42]: There is no testnet at all? Can you also include a reference to where it's designated `1` in the ecosystem?

**Review by oed — CHANGES_REQUESTED** [review]

File: arweave/caip122.md
Comment (ropats16) [reply arweave/caip122.md:28]: Do you mean something like Solana has specified on this line here?

https://github.com/ceramicnetwork/cacao/blob/d18d3f40bddd07d7a4ee5959f003c3416c9e1442/src/cacao.ts#L38

File: arweave/caip2.md
Comment (ropats16) [reply arweave/caip2.md:42]: There is no official test net yet. The designation of the main network as 1 is a value referenced from the EVM compatible ecosystem. As this kind of standard was not needed before this particular use case, there is no document I can point to for the same. This is essentially the first iteration of such a standard for the arweave ecosystem.

File: arweave/caip122.md
Comment (oed) [reply arweave/caip122.md:28]: Yes exactly.

File: arweave/caip2.md
Comment (oed) [reply arweave/caip2.md:42]: In ethereum the number has a specific meaning and all network id's are maintained on a list: https://chainlist.org/

Are there unofficial test nets? If yes, how does the client software differentiate between them?

Comment (ropats16) [reply arweave/caip2.md:42]: The Arweave protocol doesn't have built in support for multiple networks at the protocol level. HTTPS request are made to nodes but there's a single official Arweave network so there isn't a need to get a response from the nodes as to which network they are on. In case this is revised in the future, adequate documentation and PRs for the same would be submitted.

**Comment by ropats16** [comment]
@oed the latest updates specifies the type of the signature as it would be represented in the cacao block. Is this what you were looking for? And if so are there any other changes?

File: arweave/caip2.md
Comment (oed) [reply arweave/caip2.md:42]: From a quick google it seems like there are indeed testnets. On a protocol level they need to be distinguished. Also if we arbitrarily pick a number here it will fall on CASA to maintain a list of numbers for arweave chains which we definitely don't want to do!

Comment (DanMacDonald) [reply arweave/caip2.md:42]: There are testnets but they aren't actually networks of nodes so there's no need to distinguish them at a the protocol level. They aren't gossping data/transactions  on an open network.  These "testnets" expose Arweave like HTTP endpoints but are hosted as a standalone deployments.  When a developer wants to test their application they point their code at  the URL provided by the testnet and communicate directly with it. Testing works like this on arweave, in part, because the arweave protocol does not have a way to differentiate "network" at the protocol level. 

Development on Arweave is typically done using https://github.com/textury/arlocal, a shim that mocks up the arweave interfaces and mimics its behavior. The "testnets" that exist are a slightly modified version of this code that enable incoming connections to be managed by a load balancer. Example of this would be Redstone's "testnet" https://testnet.redstone.tools/

You may also find references to Testweave-sdk or Testweave-docker. Testweave coexisted with `arlocal` for some time and similarly enabled testing on the developers local machine. Instead of a node/SQLite mockup testweave ran an Arweave node in a container. Arweave node software really wasn't meant to run in this configuration and as a result it had constant bugs and inconsistencies with mainnet. Developers abandoned testweave for the better supported `arlocal` and it is now effectively deprecated with no active maintainer.

Comment (bumblefudge) [reply arweave/caip2.md:42]: is there any canonical "chainID" equivalent for these testnets, though, canonically maintained in any kind of registry on github or otherwise, as Joel was asking? would "testnet.redstone.tools" be a meaningful testnet name, on which someone might want to test a system by addressing and fetching state?

this information looks pretty useful to first-time arweave devs, would it make sense to put it into the README somewhere as an informational aside?

Comment (joshbenaron) [reply arweave/caip2.md:42]: @DanMacDonald Arweave is releasing a testnet for 2.6. Isn't this going to be its own chain?

Comment (DanMacDonald) [reply arweave/caip2.md:42]: That's true, there is a testnet for 2.6, purely for miners. I'm not sure exactly how they are accomplishing this. My guess is though a whitelist of 2.6 peers.

Comment (ropats16) [reply arweave/caip2.md:42]: Hey @oed @bumblefudge, based on the information provided by @DanMacDonald, how can we move forward? As the testnet for 2.6 isn't released yet, ans it's working mechanism is still vague, is it possible to move forth with considering the existence of only the Arweave main net at the moment and once the test net is released, the chain resolution can be revised and updated accordingly in the namespaces as well?

Comment (oed) [reply arweave/caip2.md:42]: Can we maybe use the genesis has as we do for other chains? That way if in the future Arweave introduces a testnet we don't have to maintain a list of numbers.

Comment (bumblefudge) [reply arweave/caip2.md:42]: this makes sense to me-- calling it `1` or `main` is arbitrary if the protocol doesn't internally have the concept of a chain ID and the community itself doesn't use that convention.  i worry that, as Joel rightly points out, the community might proceed happily for years without testnets or forks, and at such a time as they adopt the concept, they might number from 0 or use strings and call it `main-net` or `mainnet` instead of `main`.  For this reason, I second the recommendation for the "hash of genesis block" approach taken by many other caip-2 profiles since current protocol does seem to have the notion of "block 0" and addressing blocks by hashes.  

from [here](https://viewblock.io/arweave/block/0) I take it that the genesis hash is numbered 0, not 1, and hash-addressable as `7wIU7KolICAjClMlcZ38LZzshhI7xGkm2tDCJR7Wvhe3ESUo2-Z4-y0x1uaglRJE` 
. from [this documentation] i am inferring it that since block 422250 the format of RPC calls changed, but the block hashes still seem at first glance to be computed the same way and would fit the same regex: 
`JbtpfyJ6Kh22vciZJpZAVorejqDRB6ekbZC0QF_zOFbACpcae6L6PZsuy0gKRpqk` . 
It's interesting that the RPC doesn't include a request for block by block number, just by hash, but at least you can confirm you're on the intended chain (or at least, a fork of it...) by calling 
[`GET https://arweave.net/block/hash/7wIU7KolICAjClMlcZ38LZzshhI7xGkm2tDCJR7Wvhe3ESUo2-Z4-y0x1uaglRJE`](https://docs.arweave.org/developers/server/http-api#blocks) and checking the `indep_hash` top-level property of the result object to confirm.  It seems like that flow still works after the info RPC changes.

Anyways if this works for everyone, feel free to update the CAIP-2 profile and we can go from there-- tag me i you'd like any help.  To limit the length of CAIP2 and CAIP10 URNs to fit inside, say, HTTP headers, we have historically just substringed the first 16 or 32 characters-- the full 64 would make for unworkably long CAIP-10s and CAIP-19s.

Comment (oed) [reply arweave/caip2.md:42]: A trunkated hash of genesis might be alright as well. So the caip2 doesn't become so long. Maybe first 4 characters?

Comment (ropats16) [reply arweave/caip2.md]: Hey @bumblefudge @oed, just updated the PR to identify the network using a truncated version (first 16 characters) of the [hash of the genesis block](https://viewblock.io/arweave/block/0) in both `CAIP2` and `CAIP10` files.

**Comment by bumblefudge** [comment]
Chain ID stuff looks great-- deferring to the Spruce or Ceramic teams for confirmation that the signature suite stuff is encoded conformantly, will merge after that.

File: arweave/caip2.md
Comment (oed) [inline arweave/caip2.md:25]: I would make it even shorter! 4 characters should be plenty. Fairly low risk of collision here.

**Review by oed — APPROVED** [review]
Alright by me. Made one optional suggestion.

File: arweave/caip2.md
Comment (ropats16) [reply arweave/caip2.md]: Hey @oed just made the suggested changes.

**Review by oed — APPROVED** [review]
lgtm

**Comment by ropats16** [comment]
@oed curious over who's supposed to merge it? Don't think I have the permissions for it.

**Comment by oed** [comment]
It will have to be approved in the next CASA meeting: https://github.com/ChainAgnostic/CASA#regular-meetings

**Pull request merged**
